File: app/src/main/java/xyz/klinker/messenger/fragment/MediaGridFragment.java
package xyz.klinker.messenger.fragment;

import android.app.Fragment;
import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.VisibleForTesting;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import xyz.klinker.messenger.R;
import xyz.klinker.messenger.activity.ImageViewerActivity;
import xyz.klinker.messenger.adapter.MediaGridAdapter;
import xyz.klinker.messenger.shared.activity.AbstractSettingsActivity;
import xyz.klinker.messenger.shared.data.DataSource;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.data.model.Conversation;
import xyz.klinker.messenger.shared.data.model.Message;
import xyz.klinker.messenger.shared.util.ActivityUtils;
import xyz.klinker.messenger.shared.util.listener.MediaSelectedListener;

public class MediaGridFragment extends Fragment implements MediaSelectedListener {

    private static final String ARG_CONVERSATION_ID = "conversation_id";

    public Conversation conversation;
    private List<Message> messages;

    public static MediaGridFragment newInstance(long conversationId) {
        MediaGridFragment fragment = new MediaGridFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_CONVERSATION_ID, conversationId);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        loadConversation();
        setUpToolbar();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_media_grid, container, false);
        RecyclerView recyclerView = (RecyclerView) root.findViewById(R.id.recycler_view);

        recyclerView.setLayoutManager(new GridLayoutManager(getActivity(),
                getResources().getInteger(R.integer.images_column_count)));
        recyclerView.setAdapter(new MediaGridAdapter(messages, this));

        return root;
    }

    private void loadConversation() {
        DataSource source = getDataSource();
        conversation = source.getConversation(getActivity(), getArguments().getLong(ARG_CONVERSATION_ID));
        messages = source.getMediaMessages(getActivity(), conversation.id);
    }

    @VisibleForTesting
    DataSource getDataSource() {
        return DataSource.INSTANCE;
    }

    private void setUpToolbar() {
        getActivity().setTitle(conversation.title);

        Settings settings = Settings.get(getActivity());
        Toolbar toolbar = ((AbstractSettingsActivity) getActivity()).getToolbar();

        if (toolbar != null) {
            if (settings.useGlobalThemeColor) {
                toolbar.setBackgroundColor(settings.mainColorSet.color);
                ActivityUtils.setStatusBarColor(getActivity(), settings.mainColorSet.colorDark);
            } else {
                toolbar.setBackgroundColor(conversation.colors.color);
                ActivityUtils.setStatusBarColor(getActivity(), conversation.colors.colorDark);
            }
        }
    }


    @Override
    public void onSelected(List<Message> messageList, int selectedPosition) {
        Intent intent = new Intent(getActivity(), ImageViewerActivity.class);
        intent.putExtra(ImageViewerActivity.EXTRA_CONVERSATION_ID, getArguments().getLong(ARG_CONVERSATION_ID));
        intent.putExtra(ImageViewerActivity.EXTRA_MESSAGE_ID, messageList.get(selectedPosition).id);
        startActivity(intent);
    }
}

File: app/src/main/java/xyz/klinker/messenger/fragment/MediaGridFragment.kt
package xyz.klinker.messenger.fragment

import android.app.Activity
import android.app.Fragment
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.afollestad.dragselectrecyclerview.DragSelectRecyclerView
import xyz.klinker.messenger.R
import xyz.klinker.messenger.activity.ImageViewerActivity
import xyz.klinker.messenger.activity.MediaGridActivity
import xyz.klinker.messenger.adapter.MediaGridAdapter
import xyz.klinker.messenger.shared.activity.AbstractSettingsActivity
import xyz.klinker.messenger.shared.data.DataSource
import xyz.klinker.messenger.shared.data.FeatureFlags
import xyz.klinker.messenger.shared.data.MediaMessage
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.model.Conversation
import xyz.klinker.messenger.shared.data.model.Message
import xyz.klinker.messenger.shared.util.ActivityUtils
import xyz.klinker.messenger.shared.util.listener.MediaSelectedListener

class MediaGridFragment : Fragment(), MediaSelectedListener {

    private val fragmentActivity: Activity? by lazy { activity }

    val conversation: Conversation? by lazy { DataSource.getConversation(fragmentActivity!!, arguments.getLong(ARG_CONVERSATION_ID)) }
    private val messages: List<Message> by lazy { DataSource.getMediaMessages(fragmentActivity!!, conversation!!.id) }

    private var selectIsActive = false
    private var recyclerView: DragSelectRecyclerView? = null
    private var adapter: MediaGridAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setUpToolbar()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_media_grid, container, false)
        recyclerView = root.findViewById<View>(R.id.recycler_view) as DragSelectRecyclerView
        adapter = MediaGridAdapter(messages, this)

        recyclerView?.layoutManager = GridLayoutManager(fragmentActivity, resources.getInteger(R.integer.images_column_count))
        recyclerView?.adapter = adapter!!

        return root
    }

    private fun setUpToolbar() {
        fragmentActivity?.title = conversation!!.title

        val toolbar = (fragmentActivity as AbstractSettingsActivity).toolbar
        if (Settings.useGlobalThemeColor) {
            toolbar?.setBackgroundColor(Settings.mainColorSet.color)
            ActivityUtils.setStatusBarColor(fragmentActivity, Settings.mainColorSet.colorDark)
        } else {
            toolbar?.setBackgroundColor(conversation!!.colors.color)
            ActivityUtils.setStatusBarColor(fragmentActivity, conversation!!.colors.colorDark)
        }
    }

    override fun onSelected(messageList: List<MediaMessage>, selectedPosition: Int) {
        if (selectIsActive) {
            adapter?.setSelected(selectedPosition, !messageList[selectedPosition].selected)

            if (!messageList.any { it.selected }) {
                activateSelectMode(false)
            }
        } else {
            val intent = Intent(fragmentActivity, ImageViewerActivity::class.java)
            intent.putExtra(ImageViewerActivity.EXTRA_CONVERSATION_ID, arguments.getLong(ARG_CONVERSATION_ID))
            intent.putExtra(ImageViewerActivity.EXTRA_MESSAGE_ID, messageList[selectedPosition].message.id)
            startActivity(intent)
        }
    }

    override fun onStartDrag(index: Int) {
        if (FeatureFlags.MULTI_SELECT_MEDIA) {
            activateSelectMode(true)
            recyclerView?.setDragSelectActive(true, index)
        }
    }

    fun destroyActionMode() {
        adapter?.messages?.forEach { it.selected = false }
        adapter?.notifyDataSetChanged()
    }

    fun deleteSelected() {
        adapter?.messages
                ?.filter { it.selected }
                ?.forEach { DataSource.deleteMessage(fragmentActivity!!, it.message.id) }

        if (adapter != null && adapter?.messages != null) {
            adapter?.messages = adapter?.messages?.filter { !it.selected }!!
            adapter?.notifyDataSetChanged()
        }

        selectIsActive = false
    }

    private fun activateSelectMode(activate: Boolean) {
        selectIsActive = activate

        if (activate) {
            (fragmentActivity as MediaGridActivity).startMultiSelect()
        } else {
            (fragmentActivity as MediaGridActivity).stopMultiSelect()
            destroyActionMode()
        }
    }

    companion object {
        private val ARG_CONVERSATION_ID = "conversation_id"

        fun newInstance(conversationId: Long): MediaGridFragment {
            val fragment = MediaGridFragment()
            val args = Bundle()
            args.putLong(ARG_CONVERSATION_ID, conversationId)
            fragment.arguments = args

            return fragment
        }
    }
}
